beacon-chain/blockchain/block_processing.go

```diff
@@ -0,0 +1,194 @@
+package blockchain
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+	b "github.com/prysmaticlabs/prysm/beacon-chain/core/blocks"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/validators"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// ReceiveBlock is a function that defines the operations that are preformed on
+// any block that is received from p2p layer or rpc. It checks the block to see
+// if it passes the pre-processing conditions, if it does then the per slot
+// state transition function is carried out on the block.
+// spec:
+//  def process_block(block):
+//      if not block_pre_processing_conditions(block):
+//          return nil, error
+//
+//  	# process skipped slots
+//
+// 		while (state.slot < block.slot - 1):
+//      	state = slot_state_transition(state, block=None)
+//
+//		# process slot with block
+//		state = slot_state_transition(state, block)
+//
+//		# check state root
+//      if block.state_root == hash(state):
+//			return state, error
+//		else:
+//			return nil, error  # or throw or whatever
+//
+func (c *ChainService) ReceiveBlock(block *pb.BeaconBlock) (*pb.BeaconState, error) {
```

Comment:
Please pass in ctx, create new span, provide this ctx to beaconDB.State(ctx) and other methods. It would be incredibly useful to see how long these processes are taking.